rust/xaynet-server/src/settings.rs

```diff
@@ -558,6 +564,181 @@ where
+    /// The bucket name in which the global models are stored.
+    /// Defaults to `global-models`.
+    ///
+    /// Please follow the [rules for bucket naming](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)
+    /// when creating the name.
+    ///
+    /// # Examples
+    ///
+    /// **TOML**
+    /// ```text
+    /// [s3.buckets]
+    /// global_models = "global-models"
+    /// ```
+    ///
+    /// **Environment variable**
+    /// ```text
+    /// XAYNET_S3__BUCKETS__GLOBAL_MODELS="global-models"
+    /// ```
+    #[validate(custom = "validate_s3_bucket_name")]
+    global_models: String,
+}
+
+// Default value for the global models bucket
+impl Default for S3BucketsSettings {
+    fn default() -> Self {
+        Self {
+            global_models: String::from("global-models"),
+        }
+    }
+}
+
+// Validates the bucket name
+// [Rules for AWS bucket naming](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)
+fn validate_s3_bucket_name(bucket_name: &str) -> Result<(), ValidationError> {
+    // https://stackoverflow.com/questions/50480924/regex-for-s3-bucket-name#comment104807676_58248645
+    // I had to use fancy_regex here because the std regex does not support `look-around`
+    let re =
+        Regex::new(r"(?!^(\d{1,3}\.){3}\d{1,3}$)(^[a-z0-9]([a-z0-9-]*(\.[a-z0-9])?)*$(?<!\-))")
+            .unwrap();
+    match re.is_match(bucket_name) {
+        Ok(true) => Ok(()),
+        Ok(false) => Err(ValidationError::new("invalid bucket name\n See here: https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html")),
+        // something went wrong with the regex engine
+        Err(_) => Err(ValidationError::new("can not validate bucket name")),
+    }
+}
+
+// A small wrapper to support the list type for environment variable values.
+// config-rs always converts a environment variable value to a string
+// https://github.com/mehcode/config-rs/blob/master/src/env.rs#L114 .
+// Strings however, are not supported by the deserializer of rusoto_core::Region (only sequences).
+// Therefore we use S3RegionVisitor to implement `visit_str` and thus support
+// the deserialization of rusoto_core::Region from strings.
+fn deserialize_s3_region<'de, D>(deserializer: D) -> Result<Region, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    struct S3RegionVisitor;
+
+    impl<'de> Visitor<'de> for S3RegionVisitor {
+        type Value = Region;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("sequence of \"name Optional<endpoint>\"")
+        }
+
+        // FIXME: a copy of https://rusoto.github.io/rusoto/src/rusoto_core/region.rs.html#221-251\
+        // I haven't managed to create a sequence and call `self.visit_seq(seq)`.
+        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            let mut seq = value.split_whitespace();
```

Comment:
Could we just `collect` the iter to create a seq and then call `self.visit_seq(seq)` ?
Something like:

```rust
        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            let mut seq = value.split_whitespace().collect();
            self.visit_seq(seq)
        }
```

Reply:
Yes, that was my plan. The issue is that the deserializer of `Region` expects a sequence of `(String, Optional<String>)` so unfortunately I can't use a `vec` here. I tried a tuple, but I wasn't able to create a sequence from a tuple.

Reply:
This seems to compile:

```rust
        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            let seq = value.split_whitespace();
            self.visit_seq(serde::de::value::SeqDeserializer::new(seq))
        }
```

Reply:
Ah but the tests fail:

```
---- settings::tests::test_s3_custom_region_env stdout ----
thread 'settings::tests::test_s3_custom_region_env' panicked at 'called `Result::unwrap()` on an `Err` value: invalid type: string "http://localhost:8000", expected option', xaynet-server/src/settings.rs:1152:57
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- settings::tests::test_s3_custom_region_toml stdout ----
thread 'settings::tests::test_s3_custom_region_toml' panicked at 'called `Result::unwrap()` on an `Err` value: invalid type: string "http://localhost:8000", expected option', xaynet-server/src/settings.rs:1107:57
```

Reply:
Ok let's forget about this. I don't think we can do better.